File: ICSharpCode.Decompiler.Tests/TestCases/VBPretty/VBYieldInTryCatch.cs
using System;
using System.Collections.Generic;

using Microsoft.VisualBasic.CompilerServices;

public class VBYieldInTryCatch
{
	private sealed class ThrowingSequence
	{
		private readonly string label;

		private readonly bool throwBeforeFirstValue;

		private readonly bool throwAfterFirstValue;

		public ThrowingSequence(string label, bool throwBeforeFirstValue, bool throwAfterFirstValue)
		{
			this.label = label;
			this.throwBeforeFirstValue = throwBeforeFirstValue;
			this.throwAfterFirstValue = throwAfterFirstValue;
		}

		public ThrowingEnumerator GetEnumerator()
		{
			Console.WriteLine(label + ":get-enumerator");
			if (throwBeforeFirstValue)
			{
				throw new InvalidOperationException("before");
			}
			return new ThrowingEnumerator(label, throwAfterFirstValue);
		}
	}

	private struct ThrowingEnumerator
	{
		private readonly string label;

		private readonly bool throwAfterFirstValue;

		private int position;

		public int Current => position;

		public ThrowingEnumerator(string label, bool throwAfterFirstValue)
		{
			this = default(ThrowingEnumerator);
			this.label = label;
			this.throwAfterFirstValue = throwAfterFirstValue;
			position = 0;
		}

		public bool MoveNext()
		{
			checked
			{
				position++;
				Console.WriteLine(label + ":move-next:" + Conversions.ToString(position));
				if (throwAfterFirstValue && position == 2)
				{
					throw new InvalidOperationException("after");
				}
				return position <= 2;
			}
		}
	}

	private static IEnumerable<int> Enumerate(string label, ThrowingSequence sequence)
	{
		ThrowingEnumerator throwingEnumerator = default(ThrowingEnumerator);
		bool flag = false;
		bool flag2 = false;
		while (!flag2)
		{
			bool flag3 = false;
			int num = default(int);
			try
			{
				if (!flag)
				{
					throwingEnumerator = sequence.GetEnumerator();
					flag = true;
				}
				if (throwingEnumerator.MoveNext())
				{
					int item = throwingEnumerator.Current;
					num = TraceValue(label, item);
					flag3 = true;
				}
				else
				{
					flag2 = true;
				}
			}
			catch (InvalidOperationException ex)
			{
				ProjectData.SetProjectError(ex);
				InvalidOperationException ex2 = ex;
				Console.WriteLine(label + ":catch:" + ex2.Message);
				ProjectData.ClearProjectError();
				flag2 = true;
			}
			if (flag3)
			{
				yield return num;
			}
		}
	}

	private static int TraceValue(string label, int value)
	{
		Console.WriteLine(label + ":yield:" + Conversions.ToString(value));
		return value;
	}

	private static void Run(string label, ThrowingSequence sequence)
	{
		Console.WriteLine(label + ":start");
		foreach (int item in Enumerate(label, sequence))
		{
			Console.WriteLine(label + ":consumer:" + Conversions.ToString(item));
		}
		Console.WriteLine(label + ":end");
	}

	[STAThread]
	public static void Main()
	{
		Run("before", new ThrowingSequence("before", throwBeforeFirstValue: true, throwAfterFirstValue: false));
		Run("after", new ThrowingSequence("after", throwBeforeFirstValue: false, throwAfterFirstValue: true));
		Run("complete", new ThrowingSequence("complete", throwBeforeFirstValue: false, throwAfterFirstValue: false));
	}
}

File: ICSharpCode.Decompiler.Tests/TestCases/VBPretty/VBYieldInTryCatch.vb
Imports System
Imports System.Collections.Generic

Public Class VBYieldInTryCatch
	Private NotInheritable Class ThrowingSequence
		Private ReadOnly label As String
		Private ReadOnly throwBeforeFirstValue As Boolean
		Private ReadOnly throwAfterFirstValue As Boolean

		Public Sub New(label As String, throwBeforeFirstValue As Boolean, throwAfterFirstValue As Boolean)
			Me.label = label
			Me.throwBeforeFirstValue = throwBeforeFirstValue
			Me.throwAfterFirstValue = throwAfterFirstValue
		End Sub

		Public Function GetEnumerator() As ThrowingEnumerator
			Console.WriteLine(label & ":get-enumerator")
			If throwBeforeFirstValue Then
				Throw New InvalidOperationException("before")
			End If
			Return New ThrowingEnumerator(label, throwAfterFirstValue)
		End Function
	End Class

	Private Structure ThrowingEnumerator
		Private ReadOnly label As String
		Private ReadOnly throwAfterFirstValue As Boolean
		Private position As Integer

		Public Sub New(label As String, throwAfterFirstValue As Boolean)
			Me.label = label
			Me.throwAfterFirstValue = throwAfterFirstValue
			position = 0
		End Sub

		Public ReadOnly Property Current As Integer
			Get
				Return position
			End Get
		End Property

		Public Function MoveNext() As Boolean
			position += 1
			Console.WriteLine(label & ":move-next:" & position)
			If throwAfterFirstValue AndAlso position = 2 Then
				Throw New InvalidOperationException("after")
			End If
			Return position <= 2
		End Function
	End Structure

	Private Shared Iterator Function Enumerate(label As String, sequence As ThrowingSequence) As IEnumerable(Of Integer)
		Try
			For Each item In sequence
				Yield TraceValue(label, item)
			Next
		Catch ex As InvalidOperationException
			Console.WriteLine(label & ":catch:" & ex.Message)
		End Try
	End Function

	Private Shared Function TraceValue(label As String, value As Integer) As Integer
		Console.WriteLine(label & ":yield:" & value)
		Return value
	End Function

	Private Shared Sub Run(label As String, sequence As ThrowingSequence)
		Console.WriteLine(label & ":start")
		For Each value In Enumerate(label, sequence)
			Console.WriteLine(label & ":consumer:" & value)
		Next
		Console.WriteLine(label & ":end")
	End Sub

	Public Shared Sub Main()
		Run("before", New ThrowingSequence("before", True, False))
		Run("after", New ThrowingSequence("after", False, True))
		Run("complete", New ThrowingSequence("complete", False, False))
	End Sub
End Class
